src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -206,9 +206,20 @@ public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOExc
         }
     }
 
+    class ApplySettings implements NodeSettingsService.Listener {
+        @Override
+        public void onRefreshSettings(Settings settings) {
+            // Because the GroovyScriptEngineService knows nothing about the
+            // cache, we need to clear it here if the setting changes
+            if (settings.get(GroovyScriptEngineService.GROOVY_SCRIPT_BLACKLIST_PATCH) != null) {
+                ScriptService.this.clearCache();
```

Comment:
this will cause the cache to be cleared on each settings refresh, even unrelated ones (once the path one is set), I think that the place where we compare the current patch vs. the new one and if there is a diff, we should clear the cache.